index.coffee

```diff
@@ -148,20 +35,171 @@ onProgressHandler = (onProgressPromise, fallbackOnProgress) ->
+						downloadedPercentage = 100
+						extractedPercentage = 100
+					else
+						downloadedPercentage = downloadProgressTracker.getProgress()
+						extractedPercentage = extractionProgressTracker.getProgress()
+
+					percentage = (downloadedPercentage + extractedPercentage) // 2
+					percentage = lastPercentage = Math.max(percentage, lastPercentage)
+
+					onProgress _.merge evt, {
+						percentage
+						downloadedPercentage
+						extractedPercentage
+						totalProgress: progressRenderer(percentage)
+					}
+				catch err
+					console.warn('Progress error:', err.message ? err)
+
+	# Create a stream that transforms `docker.modem.followProgress` onProgress
+	# events to include total progress metrics.
+	pushProgress: (image, onProgress) ->
+		progressRenderer = @renderProgress
+		progressTracker = new ProgressTracker(100 * 1024) # 100 KB
+		lastPercentage = 0
+		Promise.try ->
+			return (evt) ->
+				try
+					{ id, status } = evt
+
+					pushMatch = /Image (.*) already pushed/.exec(status)
+
+					id ?= pushMatch?[1]
+					status ?= ''
+
+					if status is 'Preparing'
+						progressTracker.addLayer(id)
+					else if status is 'Pushing' and evt.progressDetail.current?
+						progressTracker.updateLayer(id, evt.progressDetail)
+					# registry v2 statuses
+					else if _.includes(['Pushed', 'Layer already exists', 'Image already exists'], status) or /^Mounted from /.test(status)
+						progressTracker.finishLayer(id)
+					# registry v1 statuses
+					else if pushMatch? or _.includes(['Already exists', 'Image successfully pushed'], status)
+						progressTracker.finishLayer(id)
+
+					percentage = if status.match(/^latest: digest: /) or status.match(/^Pushing tag for rev /)
+						100
+					else
+						progressTracker.getProgress()
+
+					percentage = lastPercentage = Math.max(percentage, lastPercentage)
+
+					onProgress _.merge evt, {
+						id
+						percentage
+						totalProgress: progressRenderer(percentage)
+					}
+				catch err
+					console.warn('Progress error:', err.message ? err)
 
 exports.DockerProgress = class DockerProgress
 	constructor: (dockerOpts) ->
 		if !(this instanceof DockerProgress)
 			return new DockerProgress(dockerOpts)
 
 		@docker = new Docker(dockerOpts)
+		@reporter = null
+
+	getProgressRenderer: (stepCount = DEFAULT_PROGRESS_BAR_STEP_COUNT) ->
+		return (percentage) ->
+			renderProgress(percentage, stepCount)
+
+	getProgressReporter: ->
+		return @reporter if @reporter?
+		docker = @docker
+		renderer = @getProgressRenderer()
+		@reporter = docker.versionAsync().then (res) ->
+			version = res['Version']
+			if version? and semver.lt(version, LEGACY_DOCKER_VERSION)
+				LegacyProgressReporter = require('./legacy').ProgressReporter
+				return new LegacyProgressReporter(renderer, docker)
+			else
+				return new ProgressReporter(renderer)
+
+	aggregateProgress: (count, onProgress) ->
+		renderer = @getProgressRenderer()
+		states = []
+		reporters = []
+		for i in [0...count]
```

Comment:
What do you think of somthing like
```coffee
states = _.times(count, -> percentage: 0)
reporters =  _.times count, (i) ->
	combinedProgressHandler(renderer, states, i, onProgress)
```
?

My main issue is that I dislike the fact that we haven't fully populated `states` before we start passing it into the `combinedProgressHandler` function, which feels really weird.  At the same time I don't doubt it works, just that it could be clearer/cleaner (I used lodash just because I'm familiar/comfortable with it and it'll convert between cs/js/ts easily, but a cs comprehension would work just as well)